### PR TITLE
ENH: custom metrics in ElasticsearchSink

### PR DESCRIPTION
**Pull request by chenqi0805**
*Issue #, if available:*
#253 

*Description of changes:*
- Added meters to produce desired custom metrics as listed in #253 .
- Modified BulkRetryStrategyTests to cover metrics.
- Modified ElasticsearchSinkIT to cover metrics.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/BulkRetryStrategy.java [hunk 1]
```diff
@@ -27,14 +32,23 @@
...
     public BulkRetryStrategy(final RequestFunction<BulkRequest, BulkResponse> requestFunction,
                              final BiConsumer<DocWriteRequest<?>, Throwable> logFailure,
+                             final PluginMetrics pluginMetrics,
                              final Supplier<BulkRequest> bulkRequestSupplier) {
         this.requestFunction = requestFunction;
         this.logFailure = logFailure;
+        this.pluginMetrics = pluginMetrics;
         this.bulkRequestSupplier = bulkRequestSupplier;
+
+        sentDocumentsCounter = pluginMetrics.counter(DOCUMENTS_SUCCESS);
```
Comment (AustinTag) [inline data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/BulkRetryStrategy.java:50]: So we have 2 metrics here: 1 for errors and one for successes. Do we have (or want) any notion of keeping track of how many requests failed and then succeeded on retry?

**Review by AustinTag — COMMENTED** [review]
Just a general question. In general, the metrics and their testing LGTM

File: data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/BulkRetryStrategy.java (hunk 1 above)
Comment (chenqi0805) [reply data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/BulkRetryStrategy.java:50]: Fair point. This should be a good metric to tell whether there could be too many requests for ES to handle. I think we can introduce a success-on-first-attempt counter to distinguish.

Comment (AustinTag) [reply data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/BulkRetryStrategy.java:50]: Sure, maybe just a todo

**Review by AustinTag — APPROVED** [review]

File: data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/BulkRetryStrategy.java (hunk 1 above)
Comment (chenqi0805) [reply data-prepper-plugins/elasticsearch/src/main/java/com/amazon/dataprepper/plugins/sink/elasticsearch/BulkRetryStrategy.java:52]: Done in latest commits

**Review by kowshikn — APPROVED** [review]

File: data-prepper-plugins/elasticsearch/build.gradle [hunk 2]
```diff
@@ -29,11 +29,13 @@ ext {
 
 dependencies {
     compile project(':data-prepper-api')
+    testCompile project(':data-prepper-api').sourceSets.test.output
     compile project(':data-prepper-plugins:common')
     implementation "org.elasticsearch.client:elasticsearch-rest-high-level-client:${es_version}"
     implementation "com.fasterxml.jackson.core:jackson-databind:${versionMap.jackson_databind}"
     implementation "com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:${versionMap.jackson_dataformat_yaml}"
     implementation 'javax.ws.rs:javax.ws.rs-api:2.1.1'
+    implementation "io.micrometer:micrometer-core:1.5.1"
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/elasticsearch/build.gradle:38]: Good ques. For some unknown reason, the dependency on micrometer is not identified if I do not explicitly include this. But this is not the case in PeerForwarder, need more investigation into why.

**Pull request merged**